Marlin/SdFatUtil.cpp

```diff
@@ -33,15 +33,26 @@ int SdFatUtil::FreeRam() {
   return &top - reinterpret_cast<char*>(sbrk(0));
 }
 #else  // __arm__
-extern char *__brkval;
-extern char __bss_end;
+extern "C" {
+  extern int  __bss_end;
+  extern int* __brkval;
+};
 /** Amount of free RAM
  * \return The number of free bytes.
  */
 int SdFatUtil::FreeRam() {
-  char top;
-  return __brkval ? &top - __brkval : &top - &__bss_end;
-}
+  int free_memory;
+  if (reinterpret_cast<int>(__brkval) == 0) {
+    // if no heap use from end of bss section
+    free_memory = reinterpret_cast<int>(&free_memory)
+                  - reinterpret_cast<int>(&__bss_end);
+  } else {
+    // use from top of stack to heap
+    free_memory = reinterpret_cast<int>(&free_memory)
+                  - reinterpret_cast<int>(__brkval);
+  }
+  return free_memory;
+};
 #endif  // __arm
```

Comment:
The use of reinterpret_cast is the more formal way to make integers out of addresses. It is required by the newer gcc compilers. It you are still running the old compilers, you won't see the issue.